packages/eslint-plugin/src/rules/strict-void-return.ts

```diff
@@ -0,0 +1,991 @@
+          tsutils.isTypeFlagSet(
+            type,
+            ts.TypeFlags.VoidLike |
+              ts.TypeFlags.Undefined |
+              ts.TypeFlags.Null |
+              ts.TypeFlags.Any |
+              ts.TypeFlags.Never,
+          ),
+        )
+      );
+    }
+
+    /**
+     * Finds errors in any expression node.
+     *
+     * Compares the type of the node against the contextual (expected) type.
+     *
+     * @returns `true` if the expected type was void function.
+     */
+    function checkExpressionNode(
+      node: TSESTree.Expression,
+      msgId: ErrorPlaceId,
+      data?: Record<string, string>,
+    ): boolean {
+      const tsNode = parserServices.esTreeNodeToTSNodeMap.get(node);
+      assert(ts.isExpression(tsNode));
+      const expectedType = checker.getContextualType(tsNode);
+
+      if (expectedType != null && isVoidReturningFunctionType(expectedType)) {
+        reportIfNonVoidFunction(node, msgId, data);
+        return true;
+      }
+
+      return false;
+    }
+
+    /**
+     * Finds errors in function calls.
+     *
+     * When checking arguments, we also manually figure out the argument types
+     * by iterating over all the function signatures.
+     * Thanks to this, we can find arguments like `(() => void) | (() => any)`
+     * and treat them as void too.
+     * This is done to also support checking functions like `addEventListener`
+     * which have overloads where one callback returns any.
+     *
+     * Implementation mostly based on no-misused-promises,
+     * which does this to find `(() => void) | (() => NotThenable)`
+     * and report them too.
+     */
+    function checkFunctionCallNode(
+      callNode: TSESTree.CallExpression | TSESTree.NewExpression,
+    ): void {
+      const funcName =
+        util.getNameFromExpression(sourceCode, callNode.callee) ?? 'function';
+
+      const callTsNode = parserServices.esTreeNodeToTSNodeMap.get(callNode);
+      for (const [argIdx, argNode] of callNode.arguments.entries()) {
+        if (argNode.type === AST_NODE_TYPES.SpreadElement) {
+          continue;
+        }
+
+        // Check against the contextual type first
+        if (checkExpressionNode(argNode, 'Arg', { funcName })) {
+          continue;
+        }
+
+        // Check against the types from all of the call signatures
+        if (options.considerOtherOverloads) {
+          const funcType = checker.getTypeAtLocation(callTsNode.expression);
+          const funcSignatures = tsutils
+            .unionTypeParts(funcType)
+            .flatMap(type =>
+              ts.isCallExpression(callTsNode)
+                ? type.getCallSignatures()
+                : type.getConstructSignatures(),
+            );
+          const argExpectedReturnTypes = funcSignatures
+            .map(s => s.parameters[argIdx])
+            // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition -- Indexing can return undefined
+            .filter(param => param != null)
```

Comment:
[Style] Minor nitpick, but helps get rid of the comment 😄 

```suggestion
            .filter(Boolean)
```